api/src/glfs-fops.c

```diff
@@ -516,6 +516,158 @@ pub_glfs_open(struct glfs *fs, const char *path, int flags)
     return glfd;
 }
 
+static void
+cleanup_fopat_args(struct glfs_fd *pglfd, xlator_t *subvol, int ret, loc_t *loc)
+{
+    if (loc)
+        loc_wipe(loc);
+
+    if (subvol)
+        glfs_subvol_done(pglfd->fs, subvol);
+
+    GF_REF_PUT(pglfd);
+}
+
+static xlator_t *
+setup_fopat_args(struct glfs_fd *pglfd, const char *path, gf_boolean_t follow,
+                 loc_t *loc, struct iatt *iatt)
+{
+    int ret = 0;
+    xlator_t *subvol = NULL;
+
+    GF_REF_GET(pglfd);
+
+    subvol = glfs_active_subvol(pglfd->fs);
+    if (!subvol) {
+        ret = -1;
+        errno = EIO;
+        goto out;
+    }
+
+    glfs_lock(pglfd->fs, _gf_true);
+    {
+        ret = glfs_resolve_at(pglfd->fs, subvol, pglfd->fd->inode, path, loc,
+                              iatt, follow, 0);
+    }
+    glfs_unlock(pglfd->fs);
+
+    if (ret < 0) {
+        goto out;
+    }
+
+    ret = 0;
+out:
+    if (ret < 0 && errno != ENOENT) {
+        cleanup_fopat_args(pglfd, subvol, ret, loc);
+        subvol = NULL;
+    }
+
+    return subvol;
+}
+
+GFAPI_SYMVER_PUBLIC_DEFAULT(glfs_openat, 11.0)
+struct glfs_fd *
+pub_glfs_openat(struct glfs_fd *pglfd, const char *path, int flags, mode_t mode)
+{
+    int ret = -1;
+    struct glfs_fd *glfd = NULL;
+    xlator_t *subvol = NULL;
+    loc_t loc = {
+        0,
+    };
+    dict_t *fop_attr = NULL;
+    struct iatt iatt = {0};
+    uuid_t gfid;
+    gf_boolean_t is_create = 0;
+
+    DECLARE_OLD_THIS;
+    __GLFS_ENTRY_VALIDATE_FD(pglfd, invalid_fs);
+
+    is_create = !!(flags & O_CREAT);
+    subvol = setup_fopat_args(pglfd, path, !(flags & O_NOFOLLOW), &loc, &iatt);
+    if (!subvol) {
+        goto out;
+    }
+
+    if (is_create && !loc.inode) {
+        loc.inode = inode_new(loc.parent->table);
+        if (!loc.inode) {
+            ret = -1;
+            errno = ENOMEM;
+            goto out;
+        }
+
+        gf_uuid_generate(gfid);
+        if (!fop_attr) {
+            /* create */
+            fop_attr = dict_new();
+            if (!fop_attr) {
+                ret = -1;
+                errno = ENOMEM;
+                goto out;
+            }
+        }
+        ret = dict_set_gfuuid(fop_attr, "gfid-req", gfid, true);
+        if (ret) {
+            ret = -1;
+            errno = ENOMEM;
+            goto out;
+        }
```

Comment:
Won't this code be used also for entry creation fops, like mknod, mkdir, ... ?

If so, probably it would be good to move it to another common function, but I'm ok to delay this "unification" to a patch for one of those fops that also require this code.

Reply:
I agree. It will be meaningful and visible when the common code is done with another entry creation fop.

Good to get this in and immediately work (PRs already out there) to get some entry fops to handle the common code